src/linear_algebra/tensor_algebra.lean

```diff
@@ -0,0 +1,248 @@
+  λ t, pre.rec_on t f (algebra_map _ _) (λ _ _, (+)) (λ _ _, (*))
+
+/--
+An inductively defined relation on `pre R M`
+used to force the initial algebra structure on the associated quotient.
+-/
+inductive rel : (pre R M) → (pre R M) → Prop
+-- force `of` to be linear
+| add_lin {a b : M} : rel ↑(a+b) (↑a + ↑b)
+| smul_lin {r : R} {a : M} : rel ↑(r • a) (↑r * ↑a)
+-- force `of_scalar` to be a central semiring morphism
+| add_scalar {r s : R} : rel ↑(r + s) (↑r + ↑s)
+| mul_scalar {r s : R} : rel ↑(r * s) (↑r * ↑s)
+| central_scalar {r : R} {a : pre R M} : rel (r * a) (a * r)
+-- commutative additive semigroup
+| add_assoc {a b c : pre R M} : rel (a + b + c) (a + (b + c))
+| add_comm {a b : pre R M} : rel (a + b) (b + a)
+| zero_add {a : pre R M} : rel (0 + a) a
+-- multiplicative monoid
+| mul_assoc {a b c : pre R M} : rel (a * b * c) (a * (b * c))
+| one_mul {a : pre R M} : rel (1 * a) a
+| mul_one {a : pre R M} : rel (a * 1) a
+-- distributivity
+| left_distrib {a b c : pre R M} : rel (a * (b + c)) (a * b + a * c)
+| right_distrib {a b c : pre R M} : rel ((a + b) * c) (a * c + b * c)
+-- other relations needed for semiring
+| zero_mul {a : pre R M} : rel (0 * a) 0
+| mul_zero {a : pre R M} : rel (a * 0) 0
+-- compatibility
+| add_compat_left {a b c : pre R M} : rel a b → rel (a + c) (b + c)
+| add_compat_right {a b c : pre R M} : rel a b → rel (c + a) (c + b)
+| mul_compat_left {a b c : pre R M} : rel a b → rel (a * c) (b * c)
+| mul_compat_right {a b c : pre R M} : rel a b → rel (c * a) (c * b)
+
+end tensor_algebra
+
+/--
+The tensor algebra of the module `M` over the commutative semiring `R`.
+-/
+def tensor_algebra := quot (tensor_algebra.rel R M)
+
+namespace tensor_algebra
+
+local attribute [instance]
+  pre.has_coe_module pre.has_coe_semiring pre.has_mul pre.has_add pre.has_zero
+  pre.has_one pre.has_scalar
+
+instance : semiring (tensor_algebra R M) :=
+{ add := quot.map₂ (+) (λ _ _ _, rel.add_compat_right) (λ _ _ _, rel.add_compat_left),
+  add_assoc := by { rintros ⟨⟩ ⟨⟩ ⟨⟩, exact quot.sound rel.add_assoc },
+  zero := quot.mk _ 0,
+  zero_add := by { rintro ⟨⟩, exact quot.sound rel.zero_add },
+  add_zero := begin
+    rintros ⟨⟩,
+    change quot.mk _ _ = _,
+    rw [quot.sound rel.add_comm, quot.sound rel.zero_add],
+  end,
+  add_comm := by { rintros ⟨⟩ ⟨⟩, exact quot.sound rel.add_comm },
+  mul := quot.map₂ (*) (λ _ _ _, rel.mul_compat_right) (λ _ _ _, rel.mul_compat_left),
+  mul_assoc := by { rintros ⟨⟩ ⟨⟩ ⟨⟩, exact quot.sound rel.mul_assoc },
+  one := quot.mk _ 1,
+  one_mul := by { rintros ⟨⟩, exact quot.sound rel.one_mul },
+  mul_one := by { rintros ⟨⟩, exact quot.sound rel.mul_one },
+  left_distrib := by { rintros ⟨⟩ ⟨⟩ ⟨⟩, exact quot.sound rel.left_distrib },
+  right_distrib := by { rintros ⟨⟩ ⟨⟩ ⟨⟩, exact quot.sound rel.right_distrib },
+  zero_mul := by { rintros ⟨⟩, exact quot.sound rel.zero_mul },
+  mul_zero := by { rintros ⟨⟩, exact quot.sound rel.mul_zero } }
```

Comment:
Can and should this be further shortened to the following?
```suggestion
  mul_zero := λ ⟨⟩, quot.sound rel.mul_zero }
```

Reply:
I don't think this works, but I could be wrong. Did you try it?

Reply:
Yeah, it doesn't work. Looks like this sort of pattern matching doesn't work with `quot`.